src/EditorFeatures/Core/DefinitionContextTracker.cs

```diff
@@ -0,0 +1,211 @@
+            }
+
+            var workspace = document.Project.Solution.Workspace;
+            if (!document.SupportsSemanticModel)
+            {
+                var goToDefinitionService = document.GetLanguageService<IGoToDefinitionService>();
+                var navigableItems = await goToDefinitionService.FindDefinitionsAsync(document, pointInRoslynSnapshot.Value.Position, cancellationToken).ConfigureAwait(false);
+                if (navigableItems != null)
+                {
+                    var navigationService = workspace.Services.GetService<IDocumentNavigationService>();
+
+                    var builder = new ArrayBuilder<CodeDefinitionWindowLocation>();
+                    foreach (var item in navigableItems)
+                    {
+                        if (navigationService.CanNavigateToPosition(workspace, item.Document.Id, item.SourceSpan.Start))
+                        {
+                            var text = await item.Document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                            var linePositionSpan = text.Lines.GetLinePositionSpan(item.SourceSpan);
+                            builder.Add(new CodeDefinitionWindowLocation(item.DisplayString, item.Document.FilePath, linePositionSpan.Start.Line, linePositionSpan.Start.Character));
+                        }
+                    }
+
+                    return builder.ToImmutable();
+                }
+
+                return ImmutableArray<CodeDefinitionWindowLocation>.Empty;
+            }
+            else
+            {
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var symbol = SymbolFinder.FindSymbolAtPosition(
+                    semanticModel,
+                    pointInRoslynSnapshot.Value.Position,
+                    workspace, bindLiteralsToUnderlyingType: true,
+                    cancellationToken: cancellationToken);
+
+                if (symbol == null)
+                {
+                    return ImmutableArray<CodeDefinitionWindowLocation>.Empty;
+                }
+
+                symbol = symbol.GetOriginalUnreducedDefinition();
+                var project = document.Project;
+
+                // Get the symbol back from the originating workspace
+
+                var symbolMappingService = document.Project.Solution.Workspace.Services.GetService<ISymbolMappingService>();
+                var mappingResult = await symbolMappingService.MapSymbolAsync(document, symbol, cancellationToken).ConfigureAwait(false);
+                if (mappingResult != null)
+                {
+                    symbol = mappingResult.Symbol;
+                    project = mappingResult.Project;
+                }
+
+                var solution = project.Solution;
+                var sourceDefinition = await SymbolFinder.FindSourceDefinitionAsync(mappingResult.Symbol, mappingResult.Project.Solution, cancellationToken).ConfigureAwait(false);
```

Comment:
Does MapSymbolAsync not just return a mapping result with the inputs if it can't do it's magic? Should it? The pattern seems really funky otherwise...